### ISSUE TITLE
[FEAT] add flag to remember the latest used editor configuration

### ISSUE THREAD

**Issue by marcosnils** [gabm/Satty#34]
I generally find myself changing the default color and size very often and I really wish if `satty` could just remember the latest setting used. I guess the most straight forward approach to do this would be to persist the editor config somewhere and then read it every time `satty` is called. 

For backwards compatibility purposes I was thinking about adding a new flag for this like `--persist-editor-cfg`. 

WDYT? would a PR be useful?

**Comment by gabm** [comment]
great idea! have you seen #30? We should add default colour and default size there!

would this already solve your use case?

Currently there is no such notion as "state". Satty is basically stateless as all config is passed to it at startup. The configuration file in #30 would be a way to not only describe the initial state but - when writable - also to persist state. Imho that might be the best approach.

We should avoid unpredictable options, i.e. specifying A but getting B...

**Comment by gabm** [comment]
I think the way forward will be to implement the configuration file as a way of persisting state and then use one of the toml editor crates to edit that file when the remember last state option is set

**Comment by marcosnils** [comment]
> I think the way forward will be to implement the configuration file as a way of persisting state and then use one of the toml editor crates to edit that file when the remember last state option is set

yes sure, this makes sense. How do you prefer to move forward about this? Can I help #30 to move forward somehow?

**Comment by gabm** [comment]
I thought about this again and had some specific doubts about us touching the users config file and also systems where the configuration file is immutable. So I asked the NixOS community over at Reddit [0] and the answers are very useful .. let me summarise

- we should never touch the users config file - that's sacred userland
- we should follow the xdg standards, there is a place for application state
- it would be easier for NixOS if the user config would be immutable to the program, but if it's not that wouldn't be deal breaker either

For us that boils down to:

- config and state are two different things - contrary to what i said above
- we can implement state without worrying about the config file and even before implementing #30
- the state for should reside in it's appropriate folder and can be anything - probably json


- [0] https://www.reddit.com/r/NixOS/s/sak64v9wmD

**Comment by marcosnils** [comment]
Awesome, great. I can start with a draft PR and we can continue iterating there.

**Comment by gabm** [comment]
I would prefer the following behaviour/spec


- we introduce a flag that enables satty to remember it's last state. if the flag is not specified, it shall behave as it does until now
- If the flag is specified we do:
  - try to open a json file from the xdg specified folder
  - load the state from there
  -  use the last selected tool from there
  - update the file on every change of state
  - the state includes: selected tool, selected colour, custom colour and selected size

**Comment by gabm** [comment]
> Awesome, great. I can start with a draft PR and we can continue iterating there.

great, looking forward!

**Comment by gabm** [comment]
I implemented config file handling in #38. It pushes the config into a global variable called `APP_STATE` that is a relm4 `SHARED_STATE`. I do think that this might also be a good approach for tracking the app state (selected tools, colors, sizes etc)

**Comment by RobertMueller2** [comment]
I've had a short look at this, I think we may be able to store a session state in state file. This would include something like

- selected colour, regardless of palette or custom colour, try to restore as palette colour if possible and use as custom colour if not
- fill
- round caps
- S/M/L and annotation size factor
- selected tool

I wouldn't go for the full config at this point.

Regarding restoration of the same, sometimes you may want to use last session's stuff and sometimes you may not. My first thought about this would be to provide a shortcut that loads the above state from the state file (if available).

The whole thing would be an experimental feature, and any further ideas would evolve from that.

Until such a thing is implemented, a possible workaround may be to use different config files for different purposes. My primary use case for this is bulk annotating recent scans (with redact) vs annotating screenshots.
